File: src/core/kem.ts
/**
 * @quxtech/pqc-crypto - Key Encapsulation Module (ML-KEM / CRYSTALS-Kyber)
 * ============================================================================
 * Implements NIST FIPS 203 ML-KEM for post-quantum key encapsulation.
 *
 * Supported algorithms:
 * - ML-KEM-768 (NIST Security Level 3, ~AES-192)
 * - ML-KEM-1024 (NIST Security Level 5, ~AES-256)
 * ============================================================================
 */

import { ml_kem768, ml_kem1024 } from '@noble/post-quantum/ml-kem';
import { bytesToHex, hexToBytes, randomBytes } from '@noble/hashes/utils';
import type { SecurityLevel, KeyPair, EncapsulationResult } from '../types.js';

// Algorithm mapping
const KEM_ALGORITHMS = {
  '3': ml_kem768,
  '5': ml_kem1024,
} as const;

const KEM_NAMES = {
  '3': 'ML-KEM-768',
  '5': 'ML-KEM-1024',
} as const;

/**
 * Generate a new ML-KEM key pair
 * @param securityLevel - NIST security level ('3' or '5')
 * @returns Key pair with publicKey and secretKey as Uint8Array
 */
export function generateKeyPair(securityLevel: SecurityLevel = '5'): KeyPair {
  const kem = KEM_ALGORITHMS[securityLevel];
  const seed = randomBytes(64); // ML-KEM requires 64 bytes seed
  return kem.keygen(seed);
}

/**
 * Generate a new ML-KEM key pair with hex-encoded keys
 * @param securityLevel - NIST security level ('3' or '5')
 * @returns Key pair with hex-encoded publicKey and secretKey
 */
export function generateKeyPairHex(securityLevel: SecurityLevel = '5'): { publicKey: string; secretKey: string } {
  const keyPair = generateKeyPair(securityLevel);
  return {
    publicKey: bytesToHex(keyPair.publicKey),
    secretKey: bytesToHex(keyPair.secretKey),
  };
}

/**
 * Encapsulate a shared secret using recipient's public key
 * @param recipientPublicKey - Recipient's ML-KEM public key (hex string or Uint8Array)
 * @param securityLevel - NIST security level ('3' or '5')
 * @returns Ciphertext and shared secret (both hex-encoded)
 */
export function encapsulate(
  recipientPublicKey: string | Uint8Array,
  securityLevel: SecurityLevel = '5'
): EncapsulationResult {
  const kem = KEM_ALGORITHMS[securityLevel];
  const publicKey = typeof recipientPublicKey === 'string'
    ? hexToBytes(recipientPublicKey)
    : recipientPublicKey;

  const { cipherText, sharedSecret } = kem.encapsulate(publicKey);

  return {
    ciphertext: bytesToHex(cipherText),
    sharedSecret: bytesToHex(sharedSecret),
  };
}

/**
 * Decapsulate to recover shared secret using secret key
 * @param ciphertext - Encapsulated ciphertext (hex string or Uint8Array)
 * @param secretKey - ML-KEM secret key (hex string or Uint8Array)
 * @param securityLevel - NIST security level ('3' or '5')
 * @returns Shared secret (hex-encoded)
 */
export function decapsulate(
  ciphertext: string | Uint8Array,
  secretKey: string | Uint8Array,
  securityLevel: SecurityLevel = '5'
): string {
  const kem = KEM_ALGORITHMS[securityLevel];
  const ct = typeof ciphertext === 'string' ? hexToBytes(ciphertext) : ciphertext;
  const sk = typeof secretKey === 'string' ? hexToBytes(secretKey) : secretKey;

  const sharedSecret = kem.decapsulate(ct, sk);
  return bytesToHex(sharedSecret);
}

/**
 * Get algorithm name for security level
 * @param securityLevel - NIST security level
 * @returns Algorithm name string
 */
export function getAlgorithmName(securityLevel: SecurityLevel = '5'): string {
  return KEM_NAMES[securityLevel];
}

/**
 * Get public key size in bytes for security level
 */
export function getPublicKeySize(securityLevel: SecurityLevel = '5'): number {
  return securityLevel === '5' ? 1568 : 1184;
}

/**
 * Get secret key size in bytes for security level
 */
export function getSecretKeySize(securityLevel: SecurityLevel = '5'): number {
  return securityLevel === '5' ? 3168 : 2400;
}

/**
 * Get ciphertext size in bytes for security level
 */
export function getCiphertextSize(securityLevel: SecurityLevel = '5'): number {
  return securityLevel === '5' ? 1568 : 1088;
}

/**
 * Get shared secret size in bytes (always 32 bytes)
 */
export function getSharedSecretSize(): number {
  return 32;
}

export default {
  generateKeyPair,
  generateKeyPairHex,
  encapsulate,
  decapsulate,
  getAlgorithmName,
  getPublicKeySize,
  getSecretKeySize,
  getCiphertextSize,
  getSharedSecretSize,
};

File: src/kem.rs
//! Key Encapsulation Mechanism (KEM) module
//!
//! Implements ML-KEM (CRYSTALS-Kyber) as per NIST FIPS 203.
//! - ML-KEM-768 for Security Level 3
//! - ML-KEM-1024 for Security Level 5

use crate::error::{Error, Result};
use crate::SecurityLevel;
use pqcrypto_kyber::{kyber1024, kyber768};
use pqcrypto_traits::kem::{Ciphertext, PublicKey, SecretKey, SharedSecret};
use serde::{Deserialize, Serialize};
use zeroize::{Zeroize, ZeroizeOnDrop};

/// KEM key pair containing public and secret keys
#[derive(Clone, Serialize, Deserialize)]
pub struct KemKeyPair {
    /// Public key (safe to share)
    pub public_key: Vec<u8>,
    /// Secret key (must be kept private)
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub secret_key: Vec<u8>,
    /// Security level
    pub security_level: SecurityLevel,
}

impl Zeroize for KemKeyPair {
    fn zeroize(&mut self) {
        self.secret_key.zeroize();
    }
}

impl Drop for KemKeyPair {
    fn drop(&mut self) {
        self.zeroize();
    }
}

/// Encapsulation result containing ciphertext and shared secret
#[derive(Clone, ZeroizeOnDrop)]
pub struct EncapsulationResult {
    /// Ciphertext to send to the secret key holder
    #[zeroize(skip)]
    pub ciphertext: Vec<u8>,
    /// Shared secret (same as what decapsulation will produce)
    pub shared_secret: Vec<u8>,
}

/// Generate a KEM key pair at the specified security level
///
/// # Arguments
/// * `level` - Security level (Level3 or Level5)
///
/// # Returns
/// A `KemKeyPair` containing the public and secret keys
///
/// # Example
/// ```
/// use qux_pqc::{kem, SecurityLevel};
///
/// let keys = kem::generate_keypair(SecurityLevel::Level5).unwrap();
/// assert!(!keys.public_key.is_empty());
/// ```
pub fn generate_keypair(level: SecurityLevel) -> Result<KemKeyPair> {
    match level {
        SecurityLevel::Level3 => {
            let (pk, sk) = kyber768::keypair();
            Ok(KemKeyPair {
                public_key: pk.as_bytes().to_vec(),
                secret_key: sk.as_bytes().to_vec(),
                security_level: level,
            })
        }
        SecurityLevel::Level5 => {
            let (pk, sk) = kyber1024::keypair();
            Ok(KemKeyPair {
                public_key: pk.as_bytes().to_vec(),
                secret_key: sk.as_bytes().to_vec(),
                security_level: level,
            })
        }
    }
}

/// Encapsulate a shared secret using a public key
///
/// # Arguments
/// * `public_key` - The recipient's public key
///
/// # Returns
/// Tuple of (ciphertext, shared_secret)
pub fn encapsulate(public_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
    // Try Level 5 first based on key size
    if public_key.len() == kyber1024::public_key_bytes() {
        encapsulate_with_level(public_key, SecurityLevel::Level5)
    } else if public_key.len() == kyber768::public_key_bytes() {
        encapsulate_with_level(public_key, SecurityLevel::Level3)
    } else {
        Err(Error::InvalidKeySize {
            expected: kyber1024::public_key_bytes(),
            actual: public_key.len(),
        })
    }
}

/// Encapsulate with explicit security level
pub fn encapsulate_with_level(
    public_key: &[u8],
    level: SecurityLevel,
) -> Result<(Vec<u8>, Vec<u8>)> {
    match level {
        SecurityLevel::Level3 => {
            let pk =
                kyber768::PublicKey::from_bytes(public_key).map_err(|_| Error::InvalidKeySize {
                    expected: kyber768::public_key_bytes(),
                    actual: public_key.len(),
                })?;
            let (ss, ct) = kyber768::encapsulate(&pk);
            Ok((ct.as_bytes().to_vec(), ss.as_bytes().to_vec()))
        }
        SecurityLevel::Level5 => {
            let pk = kyber1024::PublicKey::from_bytes(public_key).map_err(|_| {
                Error::InvalidKeySize {
                    expected: kyber1024::public_key_bytes(),
                    actual: public_key.len(),
                }
            })?;
            let (ss, ct) = kyber1024::encapsulate(&pk);
            Ok((ct.as_bytes().to_vec(), ss.as_bytes().to_vec()))
        }
    }
}

/// Decapsulate to recover the shared secret
///
/// # Arguments
/// * `ciphertext` - The ciphertext from encapsulation
/// * `secret_key` - The recipient's secret key
///
/// # Returns
/// The shared secret
pub fn decapsulate(ciphertext: &[u8], secret_key: &[u8]) -> Result<Vec<u8>> {
    // Determine level from secret key size
    if secret_key.len() == kyber1024::secret_key_bytes() {
        decapsulate_with_level(ciphertext, secret_key, SecurityLevel::Level5)
    } else if secret_key.len() == kyber768::secret_key_bytes() {
        decapsulate_with_level(ciphertext, secret_key, SecurityLevel::Level3)
    } else {
        Err(Error::InvalidKeySize {
            expected: kyber1024::secret_key_bytes(),
            actual: secret_key.len(),
        })
    }
}

/// Decapsulate with explicit security level
pub fn decapsulate_with_level(
    ciphertext: &[u8],
    secret_key: &[u8],
    level: SecurityLevel,
) -> Result<Vec<u8>> {
    match level {
        SecurityLevel::Level3 => {
            let sk =
                kyber768::SecretKey::from_bytes(secret_key).map_err(|_| Error::InvalidKeySize {
                    expected: kyber768::secret_key_bytes(),
                    actual: secret_key.len(),
                })?;
            let ct = kyber768::Ciphertext::from_bytes(ciphertext)
                .map_err(|_| Error::InvalidCiphertext("Invalid Level 3 ciphertext".into()))?;
            let ss = kyber768::decapsulate(&ct, &sk);
            Ok(ss.as_bytes().to_vec())
        }
        SecurityLevel::Level5 => {
            let sk = kyber1024::SecretKey::from_bytes(secret_key).map_err(|_| {
                Error::InvalidKeySize {
                    expected: kyber1024::secret_key_bytes(),
                    actual: secret_key.len(),
                }
            })?;
            let ct = kyber1024::Ciphertext::from_bytes(ciphertext)
                .map_err(|_| Error::InvalidCiphertext("Invalid Level 5 ciphertext".into()))?;
            let ss = kyber1024::decapsulate(&ct, &sk);
            Ok(ss.as_bytes().to_vec())
        }
    }
}

/// Get the public key size for a security level
pub fn public_key_size(level: SecurityLevel) -> usize {
    match level {
        SecurityLevel::Level3 => kyber768::public_key_bytes(),
        SecurityLevel::Level5 => kyber1024::public_key_bytes(),
    }
}

/// Get the secret key size for a security level
pub fn secret_key_size(level: SecurityLevel) -> usize {
    match level {
        SecurityLevel::Level3 => kyber768::secret_key_bytes(),
        SecurityLevel::Level5 => kyber1024::secret_key_bytes(),
    }
}

/// Get the ciphertext size for a security level
pub fn ciphertext_size(level: SecurityLevel) -> usize {
    match level {
        SecurityLevel::Level3 => kyber768::ciphertext_bytes(),
        SecurityLevel::Level5 => kyber1024::ciphertext_bytes(),
    }
}

/// Get the shared secret size (32 bytes for all levels)
pub fn shared_secret_size(_level: SecurityLevel) -> usize {
    32 // Always 32 bytes
}

/// Get the algorithm name for a security level
pub fn algorithm_name(level: SecurityLevel) -> &'static str {
    match level {
        SecurityLevel::Level3 => "ML-KEM-768",
        SecurityLevel::Level5 => "ML-KEM-1024",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_keypair_generation_level5() {
        let keys = generate_keypair(SecurityLevel::Level5).unwrap();
        assert_eq!(
            keys.public_key.len(),
            public_key_size(SecurityLevel::Level5)
        );
        assert_eq!(
            keys.secret_key.len(),
            secret_key_size(SecurityLevel::Level5)
        );
    }

    #[test]
    fn test_keypair_generation_level3() {
        let keys = generate_keypair(SecurityLevel::Level3).unwrap();
        assert_eq!(
            keys.public_key.len(),
            public_key_size(SecurityLevel::Level3)
        );
        assert_eq!(
            keys.secret_key.len(),
            secret_key_size(SecurityLevel::Level3)
        );
    }

    #[test]
    fn test_encapsulate_decapsulate_level5() {
        let keys = generate_keypair(SecurityLevel::Level5).unwrap();
        let (ciphertext, shared_secret) = encapsulate(&keys.public_key).unwrap();
        let recovered = decapsulate(&ciphertext, &keys.secret_key).unwrap();
        assert_eq!(shared_secret, recovered);
    }

    #[test]
    fn test_encapsulate_decapsulate_level3() {
        let keys = generate_keypair(SecurityLevel::Level3).unwrap();
        let (ciphertext, shared_secret) = encapsulate(&keys.public_key).unwrap();
        let recovered = decapsulate(&ciphertext, &keys.secret_key).unwrap();
        assert_eq!(shared_secret, recovered);
    }

    #[test]
    fn test_different_keypairs_produce_different_keys() {
        let keys1 = generate_keypair(SecurityLevel::Level5).unwrap();
        let keys2 = generate_keypair(SecurityLevel::Level5).unwrap();
        assert_ne!(keys1.public_key, keys2.public_key);
        assert_ne!(keys1.secret_key, keys2.secret_key);
    }

    #[test]
    fn test_shared_secret_size() {
        let keys = generate_keypair(SecurityLevel::Level5).unwrap();
        let (_, shared_secret) = encapsulate(&keys.public_key).unwrap();
        assert_eq!(shared_secret.len(), 32);
    }
}
